### PR TITLE
sql: support casting REGTYPE and REGFUNC to TEXT

### PR DESCRIPTION
**Pull request by mkysel**
This allows you to run queries like `'now'::regproc::text` and `'bool'::regtype::text`.

### Motivation

This change is required to support pgcli.

As a side note, if one does 5 levels of nesting (more casts) the server crashes in `projection_lifting`.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/sql/src/plan/typeconv.rs [hunk 1]
```diff
@@ -175,9 +175,15 @@ lazy_static! {
 
             // REGPROC
             (RegProc,Oid) => Implicit: CastRegProcToOid,
+            (RegProc, String) => Explicit: sql_impl_cast("(
+                SELECT DISTINCT(name) FROM mz_catalog.mz_functions WHERE oid = $1
```
Comment (mjibson) [inline src/sql/src/plan/typeconv.rs:179]: Why the `DISTINCT`?

File: src/sql/src/plan/typeconv.rs [hunk 2]
```diff
@@ -175,9 +175,15 @@ lazy_static! {
...
+            )"), // DISTINCT is required to handle overridden functions such as generate_series
 
             // REGTYPE
             (RegType,Oid) => Implicit: CastRegTypeToOid,
+            (RegType, String) => Explicit: sql_impl_cast("(
+                SELECT name FROM mz_catalog.mz_types WHERE oid = $1
```
Comment (mjibson) [inline src/sql/src/plan/typeconv.rs:185]: This doesn't handle the case where `$1` doesn't exist. Here's what I came up with:

```
            (RegType, String) => Explicit: sql_impl_cast("(
                SELECT COALESCE(t.name, v.x::pg_catalog.text)
                FROM
                    (VALUES ($1::pg_catalog.oid)) AS v(x)
                    LEFT JOIN mz_catalog.mz_types AS t ON t.oid = v.x
            )"),
```

Probably ditto for regproc.

File: test/sqllogictest/types.slt [hunk 3]
```diff
@@ -900,3 +918,13 @@ SELECT a::regtype FROM text_to_regtype ORDER BY a
...
 1082
 NULL
+
+query T
+SELECT 'date'::regtype::text::regtype
+----
+1082
+
+query T
+SELECT 'date'::regtype::text::regtype::text
```
Comment (mjibson) [inline test/sqllogictest/types.slt:928]: Also need to test the case where the regtype is an oid with no associated type, and the NULL case too. Same for regproc.

File: src/sql/src/plan/typeconv.rs [hunk 4]
```diff
@@ -175,9 +175,15 @@ lazy_static! {
 
             // REGPROC
             (RegProc,Oid) => Implicit: CastRegProcToOid,
```
Comment (mjibson) [inline src/sql/src/plan/typeconv.rs:177]: can you add a space after the comma while you are here, and to regtype below?

File: src/sql/src/plan/typeconv.rs (hunk 1 above)
Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:179]: see comment on the line below

File: src/sql/src/plan/typeconv.rs (hunk 2 above)
Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:185]: this is what you get in PG
```
martin=# select 1::regtype::text;
 text 
------
 1
(1 row)
```

This would be ideal:
```
SELECT ISNULL(name, $1::text) FROM mz_catalog.mz_types WHERE oid = $1;
ERROR:  function isnull(text, text) does not exist
```

File: test/sqllogictest/types.slt (hunk 3 above)
Comment (mkysel) [reply test/sqllogictest/types.slt:928]: I can add NULL. OID with no associated type is already handled on lines 867-890. There is no type associated with OID==1

File: test/sqllogictest/types.slt [hunk 5]
```diff
@@ -865,11 +884,15 @@ SELECT '1'::pg_catalog.regtype
 ----
 1
 
-query error CAST does not support casting from regtype to text
+query T
 SELECT '1'::regtype::text
+----
+NULL
```
Comment (mjibson) [inline test/sqllogictest/types.slt:890]: This is wrong, it should be `1`.

File: src/sql/src/plan/typeconv.rs (hunk 2 above)
Comment (mjibson) [reply src/sql/src/plan/typeconv.rs:185]: The query I posted above works.

**Comment by mkysel** [comment]
Hmm. It seems that we are not there yet. When I replace the original simple `SELECT` statement with the `COALESCE` the following query will crash the server:
```
SELECT  nsp.nspname schema_name,
        cls.relname table_name,
        att.attname column_name,
        att.atttypid::regtype::text type_name,
    att.atthasdef AS has_default,
        pg_catalog.pg_get_expr(def.adbin, def.adrelid, true) as default
FROM    pg_catalog.pg_attribute att
    INNER JOIN pg_catalog.pg_class cls
ON att.attrelid = cls.oid
    INNER JOIN pg_catalog.pg_namespace nsp
    ON cls.relnamespace = nsp.oid
    LEFT OUTER JOIN pg_attrdef def
    ON def.adrelid = att.attrelid
    AND def.adnum = att.attnum
WHERE   cls.relkind = ANY(ARRAY['r','p','f'])
  AND NOT att.attisdropped
  AND att.attnum  > 0
ORDER BY 1, 2, att.attnum;
```

It does not crash when I use the original version of it.

**Comment by mjibson** [comment]
What's the crash? Is it related to this feature? Is it a stack overflow?

File: src/sql/src/plan/typeconv.rs [hunk 6]
```diff
@@ -174,10 +174,24 @@ lazy_static! {
             (Oid, RegType) => Assignment: CastOidToRegType,
 
             // REGPROC
-            (RegProc,Oid) => Implicit: CastRegProcToOid,
+            (RegProc, Oid) => Implicit: CastRegProcToOid,
+            (RegProc, String) => Explicit: sql_impl_cast("(
+                SELECT DISTINCT(COALESCE(t.name, v.x::pg_catalog.text))
```
Comment (mjibson) [inline src/sql/src/plan/typeconv.rs:179]: The distinct here is wrong, it's supposed to error. See postgres:

```
mjibson=#  select 'generate_series'::regproc;
ERROR:  more than one function named "generate_series"
LINE 1: select 'generate_series'::regproc;
```

`generate_series` returning multiple entries is the actual bug, because it incorrectly re-uses the same oid.

Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:179]: same behavior in materialize.
```
materialize=> select 'generate_series'::regproc;
ERROR:  Evaluation error: more than one record produced in subquery
```

`distinct` is required to support this query `select oid, oid::regproc::text from (select oid from mz_catalog.mz_functions);`. which is an equivalent of PG `select oid, oid::regproc::text from (select oid from pg_catalog.pg_proc) as foo;`

Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:179]: it seems that PG does not have duplicates.
```
martin=# select cnt, oid from (select count(*) as cnt, oid from pg_catalog.pg_proc group by oid) as foo where cnt>1;
 cnt | oid 
-----+-----
(0 rows)
```

and MZ does:
```
materialize=> select cnt, oid from (select count(*) as cnt, oid from mz_catalog.mz_functions group by oid) where cnt>1;
 cnt | oid  
-----+------
   2 | 1066
(1 row)
```

Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:179]: confirming that one of the types is incorrect.
```
materialize=> select * from mz_catalog.mz_types where oid in (20,23);
  id   | oid | schema_id | name 
-------+-----+-----------+------
 s1002 |  20 |         2 | int8
 s1003 |  23 |         2 | int4
(2 rows)

materialize=> select * from mz_catalog.mz_functions where name like '%generate_series%';
  id   | oid  | schema_id |      name       |       arg_ids       | variadic_id 
-------+------+-----------+-----------------+---------------------+-------------
 s2040 | 1067 |         2 | generate_series | {s1003,s1003}       | 
 s2040 | 1069 |         2 | generate_series | {s1002,s1002}       | 
 s2040 | 1066 |         2 | generate_series | {s1002,s1002,s1002} | 
 s2040 | 1066 |         2 | generate_series | {s1003,s1003,s1003} | 
(4 rows)
```

```
martin=# select oid,typname from pg_catalog.pg_type where oid in (20, 23);
 oid | typname 
-----+---------
  20 | int8
  23 | int4
(2 rows)
martin=# select oid, proname, proargtypes from pg_catalog.pg_proc where proname like '%generate_series%' and pronargs=3;
 oid  |     proname     |  proargtypes   
------+-----------------+----------------
 1066 | generate_series | 23 23 23
 1068 | generate_series | 20 20 20
 3259 | generate_series | 1700 1700 1700
  938 | generate_series | 1114 1114 1186
  939 | generate_series | 1184 1184 1186
(5 rows)
```

I will fix this in a separate PR.

Comment (mkysel) [reply src/sql/src/plan/typeconv.rs:179]: fixed here. https://github.com/MaterializeInc/materialize/pull/9199. Once I rebase, I will remove the distinct

**Comment by mkysel** [comment]
> What's the crash? Is it related to this feature? Is it a stack overflow?

yes. Stack overflow

**Comment by mjibson** [comment]
Does it still crash with https://github.com/MaterializeInc/materialize/pull/9172 applied?

**Comment by mkysel** [comment]
> Does it still crash with #9172 applied?

#9172 fixes the crash. I can successfully run the tests in `pgcli.slt` in both debug and release.

**Comment by asenac** [comment]
We should file an issue for the fact that chaining `sql_impl_cast`-backed casts results in an exponential explosion when passing subqueries as parameters of those `sql_impl_cast`s. I don't think that is fixable atm, but worth logging it, just in case we ever switch to QGM as the representation built during query planning.

**Review by asenac — APPROVED** [review]

**Comment by mkysel** [comment]
this PR resolves one of the requirements in #8229

**Review by mjibson — APPROVED** [review]
